From 8d196aad96fdcb310b7314a06c3dc66abe0441c7 Mon Sep 17 00:00:00 2001
From: Beniamino Galvani <bgalvani@redhat.com>
Date: Thu, 9 May 2019 12:50:24 +0200
Subject: [PATCH 3/3] vpn: set STOPPED state when service disappears

When the VPN service daemon fails it sends a state-change signal to NM
and quits; however it is difficult to guarantee that the signal is
delivered to NM before the process termination and so in some cases
the signal is lost. Work around that by manually setting the STOPPED
state when the service disappears from the bus.
---
 src/vpn/nm-vpn-connection.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/vpn/nm-vpn-connection.c b/src/vpn/nm-vpn-connection.c
index 23e0d6f04..7e689d61d 100644
--- a/src/vpn/nm-vpn-connection.c
+++ b/src/vpn/nm-vpn-connection.c
@@ -2152,6 +2152,10 @@ _name_owner_changed (GObject *object,
 		 */
 		get_secrets (self, SECRETS_REQ_SYSTEM, NULL);
 	} else if (!owner && priv->service_running) {
+		if (   priv->service_state >= NM_VPN_SERVICE_STATE_STARTING
+		    && priv->service_state < NM_VPN_SERVICE_STATE_STOPPED)
+			plugin_state_changed (self, NM_VPN_SERVICE_STATE_STOPPED);
+
 		/* service went away */
 		priv->service_running = FALSE;
 		_LOGI ("VPN service disappeared");
-- 
2.21.GIT

